File: mask/core/vault.py
"""
Vault abstraction layer for Mask Privacy SDK.

Provides pluggable backends for token-to-plaintext storage:
  - MemoryVault: In-process dict (dev/testing, single-process only)
  - RedisVault: Redis-backed (production, multi-pod K8s)
  - DynamoDBVault: AWS DynamoDB-backed (AWS-native enterprises)
  - MemcachedVault: Memcached-backed (lightweight distributed cache)

The active vault is selected via the MASK_VAULT_TYPE env var.
"""

import os
import time
import logging
import hashlib
import threading
from abc import ABC, abstractmethod
from typing import Any, Optional, Tuple

from mask.core.fpe import generate_fpe_token
from mask.core.crypto import get_crypto_engine

logger = logging.getLogger("mask.vault")

# Abstract base

class BaseVault(ABC):
    """Interface every vault backend must implement."""

    @abstractmethod
    def store(self, token: str, plaintext: str, ttl_seconds: int, pt_hash: Optional[str] = None) -> None:
        """Persist a token → plaintext mapping with a TTL. Optionally save a reverse lookup hash."""

    @abstractmethod
    def get_token_by_plaintext_hash(self, pt_hash: str) -> Optional[str]:
        """Return the existing unexpired token for a given plaintext hash, or None."""

    @abstractmethod
    def retrieve(self, token: str) -> Optional[str]:
        """Return the plaintext for *token*, or None if missing/expired."""

def _hash_plaintext(plaintext: str) -> str:
    """Helper to deterministically hash plaintext for reverse lookups in distributed vaults."""
    return hashlib.sha256(plaintext.encode("utf-8")).hexdigest()


# In-memory implementation (single-process, dev / testing)

class MemoryVault(BaseVault):
    """Dict-backed vault.  Fast, but state is lost across processes. Thread-safe."""

    def __init__(self) -> None:
        # _store maps token -> (ciphertext, expiry, pt_hash_or_none)
        self._store: dict[str, Tuple[str, float, Optional[str]]] = {}
        self._reverse_store: dict[str, str] = {}
        self._lock = threading.Lock()

    def _cleanup(self) -> None:
        # Caller must hold _lock
        now = time.time()
        expired = [k for k, (_, exp, _h) in self._store.items() if now > exp]
        for k in expired:
            _, _, pt_hash = self._store[k]
            del self._store[k]
            if pt_hash and self._reverse_store.get(pt_hash) == k:
                del self._reverse_store[pt_hash]

    def store(self, token: str, plaintext: str, ttl_seconds: int, pt_hash: Optional[str] = None) -> None:
        with self._lock:
            self._cleanup()
            self._store[token] = (plaintext, time.time() + ttl_seconds, pt_hash)
            if pt_hash:
                self._reverse_store[pt_hash] = token
            
    def get_token_by_plaintext_hash(self, pt_hash: str) -> Optional[str]:
        with self._lock:
            self._cleanup()
            token = self._reverse_store.get(pt_hash)
            if token and token in self._store:
                return token
        return None

    def retrieve(self, token: str) -> Optional[str]:
        with self._lock:
            self._cleanup()
            entry = self._store.get(token)
            if entry is None:
                return None
            plaintext, exp, pt_hash = entry
            if time.time() > exp:
                del self._store[token]
                if pt_hash and self._reverse_store.get(pt_hash) == token:
                    del self._reverse_store[pt_hash]
                return None
            return plaintext

    def delete(self, token: str) -> None:
        with self._lock:
            entry = self._store.pop(token, None)
            if entry:
                _, _, pt_hash = entry
                if pt_hash and self._reverse_store.get(pt_hash) == token:
                    del self._reverse_store[pt_hash]


# Redis implementation (multi-pod, production)

class RedisVault(BaseVault):
    """Redis-backed vault for horizontally scaled deployments.

    Requires the `redis` package and a reachable Redis instance.
    Configure via:
        MASK_REDIS_URL  (default: redis://localhost:6379/0)
    """

    def __init__(self, **redis_kwargs: Any) -> None:
        try:
            import redis  # type: ignore
        except ImportError:
            raise ImportError(
                "The 'redis' package is required for RedisVault. "
                "Install it with: pip install redis"
            )
        url = os.environ.get("MASK_REDIS_URL", "redis://localhost:6379/0")
        
        # Merge url configuration with explicitly passed kwargs (like mTLS ssl_certfile)
        kwargs = dict(decode_responses=True)
        kwargs.update(redis_kwargs)
        
        self._client = redis.Redis.from_url(url, **kwargs)
        # Verify connectivity at init time so bad config fails fast
        self._client.ping()
        logger.info("RedisVault connected to %s", url)

    def store(self, token: str, plaintext: str, ttl_seconds: int, pt_hash: Optional[str] = None) -> None:
        pipe = self._client.pipeline()
        pipe.setex(f"mask:{token}", ttl_seconds, plaintext)
        if pt_hash:
            pipe.setex(f"mask-rev:{pt_hash}", ttl_seconds, token)
            # Store the hash so delete() can clean up the reverse mapping
            pipe.setex(f"mask-hash:{token}", ttl_seconds, pt_hash)
        pipe.execute()

    def get_token_by_plaintext_hash(self, pt_hash: str) -> Optional[str]:
        token = self._client.get(f"mask-rev:{pt_hash}")
        if token:
            # Verify the token hasn't expired independently
            if self._client.exists(f"mask:{token}"):
                return token
            else:
                self._client.delete(f"mask-rev:{pt_hash}")
        return None

    def retrieve(self, token: str) -> Optional[str]:
        return self._client.get(f"mask:{token}")

    def delete(self, token: str) -> None:
        pt_hash = self._client.get(f"mask-hash:{token}")
        pipe = self._client.pipeline()
        pipe.delete(f"mask:{token}")
        pipe.delete(f"mask-hash:{token}")
        if pt_hash:
            pipe.delete(f"mask-rev:{pt_hash}")
        pipe.execute()


# DynamoDB implementation (AWS-native enterprises)

class DynamoDBVault(BaseVault):
    """AWS DynamoDB-backed vault for AWS-native enterprise deployments.

    Requires the `boto3` package and valid AWS credentials.
    Configure via:
        MASK_DYNAMODB_TABLE  (default: mask-vault)
        MASK_DYNAMODB_REGION (default: us-east-1)
    """

    def __init__(self) -> None:
        try:
            import boto3  # type: ignore
        except ImportError:
            raise ImportError(
                "The 'boto3' package is required for DynamoDBVault. "
                "Install it with: pip install boto3"
            )
        region = os.environ.get("MASK_DYNAMODB_REGION", "us-east-1")
        self._table_name = os.environ.get("MASK_DYNAMODB_TABLE", "mask-vault")
        self._client = boto3.resource("dynamodb", region_name=region)
        self._table = self._client.Table(self._table_name)
        logger.info("DynamoDBVault connected to table %s in %s", self._table_name, region)

    def store(self, token: str, plaintext: str, ttl_seconds: int, pt_hash: Optional[str] = None) -> None:
        import time as _time
        item = {
            "token": f"mask:{token}",
            "plaintext": plaintext,
            "ttl": int(_time.time()) + ttl_seconds,
        }
        if pt_hash:
            item["ptr_hash"] = pt_hash
            self._table.put_item(Item={
                "token": f"mask-rev:{pt_hash}",
                "plaintext": token,
                "ttl": int(_time.time()) + ttl_seconds,
            })
        self._table.put_item(Item=item)
        
    def get_token_by_plaintext_hash(self, pt_hash: str) -> Optional[str]:
        import time as _time
        resp = self._table.get_item(Key={"token": f"mask-rev:{pt_hash}"})
        item = resp.get("Item")
        if not item:
            return None
        if int(_time.time()) > int(item.get("ttl", 0)):
            self._table.delete_item(Key={"token": f"mask-rev:{pt_hash}"})
            return None
            
        token = item.get("plaintext")
        # Verify the actual token still exists
        return token if self.retrieve(token) is not None else None

    def retrieve(self, token: str) -> Optional[str]:
        import time as _time
        resp = self._table.get_item(Key={"token": f"mask:{token}"})
        item = resp.get("Item")
        if item is None:
            return None
        if int(_time.time()) > int(item.get("ttl", 0)):
            # Use the stored ptr_hash (not the ciphertext) to clean reverse mapping
            pt_hash = item.get("ptr_hash")
            if pt_hash:
                self._table.delete_item(Key={"token": f"mask-rev:{pt_hash}"})
            self._table.delete_item(Key={"token": f"mask:{token}"})
            return None
        return item.get("plaintext")

    def delete(self, token: str) -> None:
        # Get the full item to extract the stored ptr_hash
        resp = self._table.get_item(Key={"token": f"mask:{token}"})
        item = resp.get("Item")
        if item:
            pt_hash = item.get("ptr_hash")
            if pt_hash:
                self._table.delete_item(Key={"token": f"mask-rev:{pt_hash}"})
        self._table.delete_item(Key={"token": f"mask:{token}"})


# Memcached implementation (lightweight distributed cache)

class MemcachedVault(BaseVault):
    """Memcached-backed vault as a lightweight alternative to Redis.

    Requires the `pymemcache` package and a reachable Memcached instance.
    Configure via:
        MASK_MEMCACHED_HOST (default: localhost)
        MASK_MEMCACHED_PORT (default: 11211)
    """

    def __init__(self, **memcache_kwargs: Any) -> None:
        try:
            from pymemcache.client.base import Client  # type: ignore
        except ImportError:
            raise ImportError(
                "The 'pymemcache' package is required for MemcachedVault. "
                "Install it with: pip install pymemcache"
            )
        host = os.environ.get("MASK_MEMCACHED_HOST", "localhost")
        port = int(os.environ.get("MASK_MEMCACHED_PORT", "11211"))
        
        # Support passing explicit tls_context kwargs for mTLS
        kwargs = dict(memcache_kwargs)
        if "tls_context" not in kwargs:
            # Check for generic `ssl_context` sometimes used and map it
            if "ssl_context" in kwargs:
                kwargs["tls_context"] = kwargs.pop("ssl_context")

        self._client = Client((host, port), **kwargs)
        logger.info("MemcachedVault connected to %s:%d", host, port)

    def store(self, token: str, plaintext: str, ttl_seconds: int, pt_hash: Optional[str] = None) -> None:
        self._client.set(f"mask:{token}", plaintext, expire=ttl_seconds)
        if pt_hash:
            self._client.set(f"mask-rev:{pt_hash}", token, expire=ttl_seconds)
            # Store the hash so delete() can clean up the reverse mapping
            self._client.set(f"mask-hash:{token}", pt_hash, expire=ttl_seconds)

    def get_token_by_plaintext_hash(self, pt_hash: str) -> Optional[str]:
        val = self._client.get(f"mask-rev:{pt_hash}")
        if val is None:
            return None
        token = val.decode("utf-8") if isinstance(val, bytes) else val
        # Verify primary key exists
        return token if self.retrieve(token) is not None else None

    def retrieve(self, token: str) -> Optional[str]:
        val = self._client.get(f"mask:{token}")
        if val is None:
            return None
        return val.decode("utf-8") if isinstance(val, bytes) else val

    def delete(self, token: str) -> None:
        # Look up the stored pt_hash (not the ciphertext)
        hash_val = self._client.get(f"mask-hash:{token}")
        pt_hash = hash_val.decode("utf-8") if isinstance(hash_val, bytes) else hash_val
        self._client.delete(f"mask:{token}")
        self._client.delete(f"mask-hash:{token}")
        if pt_hash:
            self._client.delete(f"mask-rev:{pt_hash}")


# Singleton accessor

_vault_instance: Optional[BaseVault] = None

DEFAULT_TTL = int(os.environ.get("MASK_VAULT_TTL", "600"))  # 10 min


def get_vault() -> BaseVault:
    """Return the configured vault singleton (lazy-init)."""
    global _vault_instance
    if _vault_instance is None:
        vault_type = os.environ.get("MASK_VAULT_TYPE", "memory").lower()
        if vault_type == "redis":
            _vault_instance = RedisVault()
        elif vault_type == "memory":
            _vault_instance = MemoryVault()
        elif vault_type == "dynamodb":
            _vault_instance = DynamoDBVault()
        elif vault_type == "memcached":
            _vault_instance = MemcachedVault()
        else:
            raise ValueError(
                f"Unknown MASK_VAULT_TYPE='{vault_type}'. "
                "Supported values: 'memory', 'redis', 'dynamodb', 'memcached'."
            )
        logger.info("Vault initialised: %s", type(_vault_instance).__name__)
    return _vault_instance


def reset_vault() -> None:
    """Reset the vault singleton.  Useful in tests."""
    global _vault_instance
    _vault_instance = None


# Public convenience API  (encode / decode)

def encode(raw_text: str, *, ttl: Optional[int] = None) -> str:
    """Tokenise *raw_text*, encrypt it, store in vault, return the FPE token.
    If *raw_text* has already been tokenised and is active, returns the existing token.
    """
    vault = get_vault()
    pt_hash = _hash_plaintext(raw_text)
    
    # 1. Deduplication check
    existing_token = vault.get_token_by_plaintext_hash(pt_hash)
    if existing_token:
        # Check if it actually exists in the primary store and hasn't expired.
        # The backend should handle this, but just to be safe.
        if vault.retrieve(existing_token) is not None:
            logger.debug("encode  %s → %s (cached)", repr(raw_text)[:20], existing_token)
            return existing_token
    
    # 2. Generate new token
    token = generate_fpe_token(raw_text)
    
    # 3. Encrypt the plaintext before it touches the vault
    crypto = get_crypto_engine()
    ciphertext = crypto.encrypt(raw_text)
    
    # 4. Store with reverse lookup hash
    vault.store(token, ciphertext, ttl or DEFAULT_TTL, pt_hash=pt_hash)
    
    logger.debug("encode  %s → %s (new)", repr(raw_text)[:20], token)
    return token


class DecodeError(Exception):
    """Raised when strict detokenisation fails for a token."""


def decode(token: str) -> str:
    """Detokenise *token* via O(1) vault lookup and decrypt it.

    This strict helper either returns plaintext or raises ``DecodeError``:

    - If the token is missing or expired, a ``DecodeError`` is raised.
    - If decryption fails, a ``DecodeError`` is raised.
    """
    ciphertext = get_vault().retrieve(token)
    if ciphertext is None:
        logger.warning("Token not found or expired: %s", token)
        raise DecodeError("Token not found or expired")

    try:
        crypto = get_crypto_engine()
        plaintext = crypto.decrypt(ciphertext)
    except Exception as exc:
        logger.error("Failed to decrypt token %s payload.", token)
        raise DecodeError("Failed to decrypt token payload") from exc

    logger.debug("decode  %s → %s", token, repr(plaintext)[:20])
    return plaintext


def _decode_lenient(token: str) -> str:
    """Internal helper used by integrations that prefer lenient semantics.

    Returns plaintext when strict ``decode`` succeeds, otherwise falls back
    to returning the original *token* string. This preserves legacy behaviour
    for callers that explicitly opt in to resilience over strictness.
    """
    try:
        return decode(token)
    except DecodeError:
        return token

File: mask-ts/src/core/vault.ts
/**
 * Vault abstraction layer for Mask Privacy SDK.
 *
 * Provides pluggable backends for token-to-plaintext storage:
 *   - MemoryVault: In-process Map (dev/testing, single-process only)
 *   - RedisVault: Redis-backed (production, multi-pod K8s)
 *   - DynamoDBVault: AWS DynamoDB-backed (AWS-native enterprises)
 *   - MemcachedVault: Memcached-backed (lightweight distributed cache)
 *
 * The active vault is selected via the MASK_VAULT_TYPE env var.
 */

import * as process from 'process';
import * as crypto from 'crypto';
import { generateFPEToken } from './fpe';
import { looksLikeToken, TOKEN_PATTERN } from './fpe_utils';
import { getCryptoEngineAsync } from './crypto';
import { MaskVaultConnectionError } from './exceptions';
import { getAuditLogger } from '../telemetry/audit_logger';
import { BucketManager } from './search';

/**
 * Return the configured fail strategy: 'open' (default) or 'closed'.
 *
 * - open:   vault errors return null / original text (graceful).
 * - closed: vault errors raise MaskVaultConnectionError (strict).
 */

/** @internal - exported for tests */
export function _getFailStrategy(): string {
  const strategy = (process.env.MASK_FAIL_STRATEGY || "open").toLowerCase();
  return strategy;
}

/** Interface every vault backend must implement. */
export abstract class BaseVault {
  /** Persist a token → plaintext mapping with a TTL. Optionally save a reverse lookup hash. */
  abstract store(token: string, plaintext: string, ttlSeconds: number, ptHash?: string | null): Promise<void>;

  /** Return the existing unexpired token for a given plaintext hash, or null. */
  abstract getTokenByPlaintextHash(ptHash: string): Promise<string | null>;

  /** Return the plaintext for token, or null if missing/expired. */
  abstract retrieve(token: string): Promise<string | null>;

  /** Delete a token and its reverse mapping. */
  abstract delete(token: string): Promise<void>;
}

/**
 * Helper to deterministically hash plaintext for reverse lookups in distributed vaults.
 *
 * If a secret is provided (from CryptoEngine.getIndexSecret), it uses HMAC-SHA256
 * for a secure blind index. Otherwise, defaults to plain SHA-256 for backward compatibility.
 */
/** @internal - exported for tests */
export function _hashPlaintext(plaintext: string, secret?: Buffer): string {
  const trimmed = plaintext.trim();
  if (secret) {
    return crypto.createHmac('sha256', secret).update(trimmed, 'utf-8').digest('hex');
  }
  return crypto.createHash('sha256').update(trimmed, 'utf-8').digest('hex');
}

/**
 * In-memory implementation (single-process, dev / testing)
 *
 * Map-backed vault. Fast, but state is lost across processes.
 */
export class MemoryVault extends BaseVault {
  private _store: Map<string, { plaintext: string; expiry: number; ptHash: string | null }>;
  private _reverseStore: Map<string, string>;

  constructor() {
    super();
    this._store = new Map();
    this._reverseStore = new Map();
  }

  private _cleanup(): void {
    // Probabilistic cleanup: only run ~1% of the time to avoid O(N) blocking
    if (Math.random() > 0.01) {
      return;
    }

    const now = Date.now() / 1000;
    for (const [token, entry] of this._store.entries()) {
      if (now > entry.expiry) {
        this._store.delete(token);
        if (entry.ptHash && this._reverseStore.get(entry.ptHash) === token) {
          this._reverseStore.delete(entry.ptHash);
        }
      }
    }
  }

  async store(token: string, plaintext: string, ttlSeconds: number, ptHash: string | null = null): Promise<void> {
    this._cleanup();
    this._store.set(token, {
      plaintext,
      expiry: (Date.now() / 1000) + ttlSeconds,
      ptHash
    });
    if (ptHash) {
      this._reverseStore.set(ptHash, token);
    }
  }

  async getTokenByPlaintextHash(ptHash: string): Promise<string | null> {
    this._cleanup();
    const token = this._reverseStore.get(ptHash);
    if (token && this._store.has(token)) {
      return token;
    }
    return null;
  }

  async retrieve(token: string): Promise<string | null> {
    this._cleanup();
    const entry = this._store.get(token);
    if (!entry) {
      return null;
    }
    if ((Date.now() / 1000) > entry.expiry) {
      this._store.delete(token);
      if (entry.ptHash && this._reverseStore.get(entry.ptHash) === token) {
        this._reverseStore.delete(entry.ptHash);
      }
      return null;
    }
    return entry.plaintext;
  }

  async delete(token: string): Promise<void> {
    const entry = this._store.get(token);
    if (entry) {
      this._store.delete(token);
      if (entry.ptHash && this._reverseStore.get(entry.ptHash) === token) {
        this._reverseStore.delete(entry.ptHash);
      }
    }
  }
}

/**
 * Redis-backed vault for horizontally scaled deployments.
 *
 * Requires the 'ioredis' package and a reachable Redis instance.
 * Configure via:
 *     MASK_REDIS_URL  (default: redis://localhost:6379/0)
 */
export class RedisVault extends BaseVault {
  private _client: any; // ioredis.Redis

  constructor(options: any = {}) {
    super();
    try {
      const Redis = require('ioredis');
      const url = process.env.MASK_REDIS_URL || "redis://localhost:6379/0";
      this._client = new Redis(url, {
        ...options,
        // Ensure connectivity at init time
        maxRetriesPerRequest: 3
      });
      console.info(`RedisVault connected to ${url}`);
    } catch (e) {
      throw new MaskVaultConnectionError(`Failed to connect to Redis: ${e}`);
    }
  }

  async store(token: string, ciphertext: string, ttlSeconds: number, ptHash: string | null = null): Promise<void> {
    try {
      const pipeline = this._client.pipeline();
      pipeline.set(`mask:${token}`, ciphertext, 'EX', ttlSeconds);
      if (ptHash) {
        pipeline.set(`mask-rev:${ptHash}`, token, 'EX', ttlSeconds);
        pipeline.set(`mask-hash:${token}`, ptHash, 'EX', ttlSeconds);
      }
      const results = await pipeline.exec();
      if (results) {
        for (const [err] of results) {
          if (err) throw err;
        }
      }
    } catch (e) {
      if (_getFailStrategy() === "closed") {
        throw new MaskVaultConnectionError(`Redis error: ${e}`);
      }
    }
  }

  async getTokenByPlaintextHash(ptHash: string): Promise<string | null> {
    const token = await this._client.get(`mask-rev:${ptHash}`);
    if (token) {
      if (await this._client.exists(`mask:${token}`)) {
        return token;
      } else {
        await this._client.del(`mask-rev:${ptHash}`);
      }
    }
    return null;
  }

  async retrieve(token: string): Promise<string | null> {
    return await this._client.get(`mask:${token}`);
  }

  async delete(token: string): Promise<void> {
    const ptHash = await this._client.get(`mask-hash:${token}`);
    const pipeline = this._client.pipeline();
    pipeline.del(`mask:${token}`);
    pipeline.del(`mask-hash:${token}`);
    if (ptHash) {
      pipeline.del(`mask-rev:${ptHash}`);
    }
    await pipeline.exec();
  }
}

/**
 * AWS DynamoDB-backed vault for AWS-native enterprise deployments.
 */
export class DynamoDBVault extends BaseVault {
  private _client: any;
  private _tableName: string;
  private _region: string;

  constructor() {
    super();
    const { DynamoDBClient } = require('@aws-sdk/client-dynamodb');
    const { DynamoDBDocument } = require('@aws-sdk/lib-dynamodb');
    this._region = process.env.MASK_DYNAMODB_REGION || "us-east-1";
    this._tableName = process.env.MASK_DYNAMODB_TABLE || "mask-vault";
    
    const baseClient = new DynamoDBClient({ region: this._region });
    this._client = DynamoDBDocument.from(baseClient);
    console.info(`DynamoDBVault connected to table ${this._tableName} in ${this._region}`);
  }

  async store(token: string, ciphertext: string, ttlSeconds: number, ptHash: string | null = null): Promise<void> {
    const { TransactWriteCommand, PutCommand } = require('@aws-sdk/lib-dynamodb');
    const now = Math.floor(Date.now() / 1000);
    const ttlVal = now + ttlSeconds;
    const item = {
      token: `mask:${token}`,
      ciphertext: ciphertext,
      ttl: ttlVal,
      ptr_hash: ptHash || undefined
    };

    if (ptHash) {
      try {
        await this._client.send(new TransactWriteCommand({
          TransactItems: [
            {
              Put: {
                TableName: this._tableName,
                Item: {
                  token: `mask:${token}`,
                  ciphertext: ciphertext,
                  ttl: ttlVal,
                  ptr_hash: ptHash
                }
              }
            },
            {
              Put: {
                TableName: this._tableName,
                Item: {
                  token: `mask-rev:${ptHash}`,
                  ciphertext: token,
                  ttl: ttlVal
                }
              }
            }
          ]
        }));
      } catch (e: any) {
        console.error(`DynamoDB transact_write_items failed: ${e}`);
        // In both strategies, we must raise for DynamoDB failures to prevent data loss
        // since we didn't perform the atomic write.
        throw new MaskVaultConnectionError(`DynamoDB atomic write failed: ${e}`);
      }
    } else {
      // Single store (no reverse index)
      try {
        await this._client.send(new PutCommand({ TableName: this._tableName, Item: item }));
      } catch (e: any) {
        if (_getFailStrategy() === "closed") {
          throw new MaskVaultConnectionError(`DynamoDB individual write failed: ${e}`);
        }
      }
    }
  }

  async getTokenByPlaintextHash(ptHash: string): Promise<string | null> {
    const { GetCommand, DeleteCommand } = require('@aws-sdk/lib-dynamodb');
    const now = Math.floor(Date.now() / 1000);
    const resp = await this._client.send(new GetCommand({
      TableName: this._tableName,
      Key: { token: `mask-rev:${ptHash}` }
    }));
    const item = resp.Item;
    if (!item) return null;
    if (now > (item.ttl || 0)) {
      await this._client.send(new DeleteCommand({ TableName: this._tableName, Key: { token: `mask-rev:${ptHash}` } }));
      return null;
    }
    const token = item.ciphertext;
    return (await this.retrieve(token)) !== null ? token : null;
  }

  async retrieve(token: string): Promise<string | null> {
    const { GetCommand, DeleteCommand } = require('@aws-sdk/lib-dynamodb');
    const now = Math.floor(Date.now() / 1000);
    const resp = await this._client.send(new GetCommand({
      TableName: this._tableName,
      Key: { token: `mask:${token}` }
    }));
    const item = resp.Item;
    if (!item) return null;
    if (now > (item.ttl || 0)) {
      const ptHash = item.ptr_hash;
      if (ptHash) {
        await this._client.send(new DeleteCommand({ TableName: this._tableName, Key: { token: `mask-rev:${ptHash}` } }));
      }
      await this._client.send(new DeleteCommand({ TableName: this._tableName, Key: { token: `mask:${token}` } }));
      return null;
    }
    return item.ciphertext;
  }

  async delete(token: string): Promise<void> {
    const { GetCommand, DeleteCommand } = require('@aws-sdk/lib-dynamodb');
    const resp = await this._client.send(new GetCommand({
      TableName: this._tableName,
      Key: { token: `mask:${token}` }
    }));
    const item = resp.Item;
    if (item && item.ptr_hash) {
      await this._client.send(new DeleteCommand({ TableName: this._tableName, Key: { token: `mask-rev:${item.ptr_hash}` } }));
    }
    await this._client.send(new DeleteCommand({ TableName: this._tableName, Key: { token: `mask:${token}` } }));
  }
}

/**
 * Memcached-backed vault.
 */
export class MemcachedVault extends BaseVault {
  private _client: any;

  constructor(options: any = {}) {
    super();
    try {
      const memjs = require('memjs');
      const host = process.env.MASK_MEMCACHED_HOST || "localhost";
      const port = process.env.MASK_MEMCACHED_PORT || "11211";
      this._client = memjs.Client.create(`${host}:${port}`, options);
      console.info(`MemcachedVault connected to ${host}:${port}`);
    } catch (e) {
      throw new MaskVaultConnectionError(`Failed to connect to Memcached: ${e}`);
    }
  }

  async store(token: string, ciphertext: string, ttlSeconds: number, ptHash: string | null = null): Promise<void> {
    try {
        await this._client.set(`mask:${token}`, Buffer.from(ciphertext), { expires: ttlSeconds });
        if (ptHash) {
           await this._client.set(`mask-rev:${ptHash}`, Buffer.from(token), { expires: ttlSeconds });
           await this._client.set(`mask-hash:${token}`, Buffer.from(ptHash), { expires: ttlSeconds });
        }
    } catch (e) {
        if (_getFailStrategy() === "closed") {
            throw new MaskVaultConnectionError(`Memcached error: ${e}`);
        }
    }
  }

  async getTokenByPlaintextHash(ptHash: string): Promise<string | null> {
    try {
        const { value } = await this._client.get(`mask-rev:${ptHash}`);
        if (!value) return null;
        const token = value.toString();
        return (await this.retrieve(token)) !== null ? token : null;
    } catch (e) {
        return null;
    }
  }

  async retrieve(token: string): Promise<string | null> {
    try {
        const { value } = await this._client.get(`mask:${token}`);
        return value ? value.toString() : null;
    } catch (e) {
        return null;
    }
  }

  async delete(token: string): Promise<void> {
    try {
        const { value } = await this._client.get(`mask-hash:${token}`);
        const ptHash = value ? value.toString() : null;
        await this._client.delete(`mask:${token}`);
        await this._client.delete(`mask-hash:${token}`);
        if (ptHash) {
          await this._client.delete(`mask-rev:${ptHash}`);
        }
    } catch (e) {}
  }
}

// Singleton accessor

let _vaultInstance: BaseVault | null = null;
const DEFAULT_TTL = parseInt(process.env.MASK_VAULT_TTL || "600");

export function getVault(): BaseVault {
  if (_vaultInstance === null) {
    const vaultType = (process.env.MASK_VAULT_TYPE || "memory").toLowerCase();
    switch (vaultType) {
      case "redis":
        _vaultInstance = new RedisVault();
        break;
      case "memory":
        _vaultInstance = new MemoryVault();
        break;
      case "dynamodb":
        _vaultInstance = new DynamoDBVault();
        break;
      case "memcached":
        _vaultInstance = new MemcachedVault();
        break;
      default:
        throw new Error(`Unknown MASK_VAULT_TYPE='${vaultType}'. Supported values: 'memory', 'redis', 'dynamodb', 'memcached'.`);
    }
    console.info(`Vault initialised: ${_vaultInstance.constructor.name}`);
  }
  return _vaultInstance;
}

export function resetVault(): void {
  _vaultInstance = null;
}

// Public convenience API (encode / decode)

export type EncodeOptions = {
    ttl?: number;
    searchBuckets?: ('year' | 'month' | 'day' | 'numeric')[];
    searchBucketSize?: number;
};

/**
 * Tokenise rawText, encrypt it, store in vault, return the FPE token.
 */
export async function encode(rawText: string, options: EncodeOptions = {}): Promise<string> {
  if (looksLikeToken(rawText)) {
    return rawText;
  }

  const text = rawText.trim();
  const vault = getVault();
  const cryptoEngine = await getCryptoEngineAsync();
  const indexSecret = await cryptoEngine.getIndexSecret();
  
  const ptHash = _hashPlaintext(text, indexSecret);

  // 1. Deduplication check
  const existingToken = await vault.getTokenByPlaintextHash(ptHash);
  if (existingToken) {
    if (await vault.retrieve(existingToken) !== null) {
      getAuditLogger().log("dedup", existingToken);
      return existingToken;
    }
  }

  // 2. Generate new token
  const token = await generateFPEToken(text);

  // 3. Encrypt the plaintext before it touches the vault
  const ciphertext = cryptoEngine.encrypt(text);

  // 4. Store with primary reverse lookup hash
  const ttl = options.ttl || DEFAULT_TTL;
  await vault.store(token, ciphertext, ttl, ptHash);

  // 5. Store additional blind indices if buckets are requested
  if (options.searchBuckets && options.searchBuckets.length > 0) {
      for (const bType of options.searchBuckets) {
          let bucketVal: string;
          if (bType === 'numeric') {
              bucketVal = BucketManager.numericBucket(text, options.searchBucketSize || 10);
          } else {
              bucketVal = BucketManager.dateBucket(text, bType);
          }
          const bHash = await BucketManager.getBucketIndex(bucketVal);
          await vault.store(token, ciphertext, ttl, bHash);

      }
  }

  getAuditLogger().log("encode", token);
  return token;
}

/** Legacy / Sync-friendly wrapper for MemoryVault only.
 * @deprecated Use encode() which is async.
 */
export function encodeSync(rawText: string, options: { ttl?: number } = {}): string {
    throw new Error("encodeSync is deprecated and unsupported in the new async architecture. Use encode().");
}

/** Async wrapper for encode (parity with Python aencode). */
export const aencode = encode;

export class DecodeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'DecodeError';
  }
}

/**
 * Detokenise token via vault lookup and decrypt it.
 */
export async function decode(token: string): Promise<string> {
  const ciphertext = await getVault().retrieve(token);
  if (ciphertext === null) {
    throw new DecodeError("Token not found or expired");
  }

  try {
    const crypto = await getCryptoEngineAsync();
    const result = crypto.decrypt(ciphertext);
    getAuditLogger().log("decode", token);
    return result;
  } catch (e) {
    throw new DecodeError("Failed to decrypt token payload");
  }
}

/** Async wrapper for decode (parity with Python adecode). */
export const adecode = decode;

/**
 * Internal helper used by integrations that prefer lenient semantics.
 */
export async function _decodeLenient(token: string): Promise<string> {
  try {
    return await decode(token);
  } catch (e) {
    return token;
  }
}

/**
 * Find all Mask tokens within text and replace them with their original plaintext.
 */
export async function detokenizeText(text: string): Promise<string> {
  if (!text || typeof text !== 'string') {
    return text;
  }

  const tokens = text.match(TOKEN_PATTERN) || [];
  let result = text;
  
  for (const token of tokens) {
      const plaintext = await _decodeLenient(token);
      if (plaintext !== token) {
        // Use global regex to replace all occurrences of this token
        // Escape the token to safely use in a regex
        const escapedToken = token.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        result = result.replace(new RegExp(escapedToken, 'g'), plaintext);
      }
  }
  
  return result;
}

/** Async wrapper for detokenizeText (parity with Python adetokenize_text). */
export const adetokenizeText = detokenizeText;
